### DISCUSSION TITLE
E2 Function Objects

### DISCUSSION THREAD

**Discussion by vurvdev** [wiremod/wire#2777] [Suggestions]
```ts
const E = function(A:number) {
    print(A)
}

const Returning = function(B:number) -> number {
    return B + 5
}

const Ret = Returning(5)[number]
const Ret = Returning("str") # Throws runtime error, invalid arguments
const Hmm = E()[string] # Throws runtime error, wrong return type

const Wrapper = function(F:function) -> function {
    return function() {
        print(F()[string])
    }
}

const Wrapped = Wrapper(function() {
    return "Hello, world!"
})[function]

Wrapped()
```

With the current type system this is how I imagine `function`s looking, essentially type erased at compile time since the type system can't handle anything as complex as functions as of now.

**Comment by Denneisk** [comment]
What about the current function syntax?

**Reply by vurvdev** [reply]
I think that's out of the scope of this sort of "RFC" and can be thought of later.

If you mean
```ts
function test() {}
const TestFn = test
```

That would require lowercase idents as variables and potentially prevent some future optimizations like function inlining and static functions, anyway just need another discussion after this is done to deal with that.

**Comment by Denneisk** [comment]
Completed
